### PR TITLE
Rbpf v0.2.0

### PR DESCRIPTION
**Pull request by Lichtso**
#### Problem

The rbpf API has changed to unify the interfaces of the interpreter and the JIT.

#### Summary of Changes

Updates the interfaces and refactors the common code of test and bench targets into the bpf_test_utils crate.
Note: The JIT is not enabled by default yet.

Fixes #

### CODE REVIEW

File: programs/bpf_loader/src/allocator_bump.rs [hunk 1]
```diff
@@ -6,35 +6,35 @@ use std::alloc::Layout;
 #[derive(Debug)]
 pub struct BPFAllocator {
     heap: Vec<u8>,
-    start: u64,
```
Comment (jackcmay) [inline programs/bpf_loader/src/allocator_bump.rs:9]: Why renamed, even if necessary doesn't below in this pr

File: runtime/src/process_instruction.rs [hunk 2]
```diff
@@ -73,6 +73,8 @@ pub trait InvokeContext {
     fn record_instruction(&self, instruction: &Instruction);
     /// Get the bank's active feature set
     fn is_feature_active(&self, feature_id: &Pubkey) -> bool;
+    /// Is JIT compilation and execution enabled
+    fn is_jit_enabled(&self) -> bool;
```
Comment (jackcmay) [inline runtime/src/process_instruction.rs:77]: Any reason not to make this a feature via featureset?

File: core/src/cluster_info.rs [hunk 3]
```diff
@@ -356,7 +356,7 @@ pub fn make_accounts_hashes_message(
 }
 
 // TODO These messages should go through the gpu pipeline for spam filtering
-#[frozen_abi(digest = "3ZHQscZ9SgxKh45idzHv3hiagyyPRtDgeySmJn171PTi")]
```
Comment (jackcmay) [inline core/src/cluster_info.rs:359]: What was the cause of this ABI change?

Comment (t-nelson) [reply core/src/cluster_info.rs:359]: I also [hit this](https://buildkite.com/solana-labs/solana/builds/33364#ebba9083-953f-4c86-a60a-a25d79f88eda/568-1211) when I "resolved" a merge conflict in `Cargo.lock` by deleting it and regenerating with `cargo tree`.  It turns out that doing so results in much broader changes than `git checkout Cargo.lock; cargo tree`.  I suspect some dependency's ABI changed and propagated to our types.

You might try resetting `Cargo.lock` to the base and regenerating it

File: programs/bpf_loader/src/lib.rs [hunk 4]
```diff
@@ -106,27 +107,36 @@ pub fn create_and_cache_executor(
     .map_err(|e| map_ebpf_error(invoke_context, EbpfError::UserError(e)))?;
     let executor = Arc::new(BPFExecutor { executable });
     invoke_context.add_executor(program.unsigned_key(), executor.clone());
+    // TODO: Cache JIT compiled code
```
Comment (jackcmay) [inline programs/bpf_loader/src/lib.rs:110]: Leave TODOs out of the code, file an issue instead

File: programs/bpf_loader/src/lib.rs [hunk 5]
```diff
@@ -258,14 +270,15 @@ impl Executor for BPFExecutor {
...
+            let mut instruction_meter = ThisInstructionMeter::new(compute_meter.clone());
             let before = compute_meter.borrow().get_remaining();
-            let result = vm.execute_program_metered(
-                parameter_bytes.as_slice(),
-                &[],
-                &[heap_region],
-                instruction_meter,
-            );
+            let result = if is_jit_enabled {
+                // TODO: Cache JIT compiled code
```
Comment (jackcmay) [inline programs/bpf_loader/src/lib.rs:276]: Remove

File: programs/bpf_loader/src/lib.rs [hunk 6]
```diff
@@ -258,14 +270,15 @@ impl Executor for BPFExecutor {
...
+                vm.jit_compile().unwrap();
+                unsafe { vm.execute_program_jit(&mut instruction_meter) }
```
Comment (jackcmay) [inline programs/bpf_loader/src/lib.rs:278]: Can you push this unsafe down into rbpf?

File: programs/bpf_loader/src/lib.rs [hunk 7]
```diff
@@ -258,14 +270,15 @@ impl Executor for BPFExecutor {
...
             let before = compute_meter.borrow().get_remaining();
-            let result = vm.execute_program_metered(
-                parameter_bytes.as_slice(),
-                &[],
-                &[heap_region],
-                instruction_meter,
-            );
+            let result = if is_jit_enabled {
+                // TODO: Cache JIT compiled code
+                vm.jit_compile().unwrap();
```
Comment (jackcmay) [inline programs/bpf_loader/src/lib.rs:277]: this goes away when the executable contains the jit compiled code?

File: programs/bpf_loader/src/lib.rs [hunk 8]
```diff
@@ -244,9 +254,11 @@ impl Executor for BPFExecutor {
         )?;
         {
             let compute_meter = invoke_context.get_compute_meter();
-            let (mut vm, heap_region) = match create_vm(
+            let is_jit_enabled = invoke_context.is_feature_active(&just_in_time_compilation::id());
+            let mut vm = match create_vm(
```
Comment (jackcmay) [inline programs/bpf_loader/src/lib.rs:258]: Side note, at this point maybe we don't have to create a vm, just call `vm::execute(..)`?

File: programs/bpf_loader/src/syscalls.rs [hunk 9]
```diff
@@ -88,147 +88,139 @@ impl SyscallConsume for Rc<RefCell<dyn ComputeMeter>> {
...
     let compute_budget = invoke_context.get_compute_budget();
 
     // Syscall functions common across languages
 
-    vm.register_syscall_ex("abort", syscall_abort)?;
-    vm.register_syscall_with_context_ex("sol_panic_", Box::new(SyscallPanic { loader_id }))?;
-    vm.register_syscall_with_context_ex(
-        "sol_log_",
-        Box::new(SyscallLog {
+    vm.register_syscall(hash_symbol_name(b"abort"), Syscall::Function(syscall_abort))?;
```
Comment (jackcmay) [inline programs/bpf_loader/src/syscalls.rs:102]: Why expose `hash_sybmol_name` up to this level?

File: programs/bpf_test_utils/Cargo.toml [hunk 10]
```diff
@@ -0,0 +1,13 @@
+[package]
+name = "solana-bpf-test-utils"
```
Comment (jackcmay) [inline programs/bpf_test_utils/Cargo.toml:2]: This could be part of the sdk to avoid releasing another crate

File: runtime/src/feature_set.rs [hunk 11]
```diff
@@ -73,6 +73,10 @@ pub mod pull_request_ping_pong_check {
     solana_sdk::declare_id!("5RzEHTnf6D7JPZCvwEzjM19kzBsyjSU3HoMfXaQmVgnZ");
 }
 
+pub mod just_in_time_compilation {
```
Comment (jackcmay) [inline runtime/src/feature_set.rs:76]: Wanna add "bpf" to the name of this feature to clarify what's being jit'd?

File: runtime/src/feature_set.rs [hunk 12]
```diff
@@ -93,6 +97,7 @@ lazy_static! {
         (cumulative_rent_related_fixes::id(), "rent fixes (#10206, #10468, #11342)"),
         (pubkey_log_syscall_enabled::id(), "pubkey log syscall"),
         (pull_request_ping_pong_check::id(), "ping-pong packet check #12794"),
+        (just_in_time_compilation::id(), "just in time compilation #12951"),
```
Comment (jackcmay) [inline runtime/src/feature_set.rs:100]: nit, hyphenate to "just-in-time compilation"

File: runtime/src/serde_snapshot.rs [hunk 13]
```diff
@@ -314,7 +314,7 @@ where
                 std::fs::rename(append_vec_abs_path.clone(), target).or_else(|_| {
                     let mut copy_options = CopyOptions::new();
                     copy_options.overwrite = true;
-                    fs_extra::move_items(&vec![&append_vec_abs_path], &local_dir, &copy_options)
+                    fs_extra::move_items(&[&append_vec_abs_path], &local_dir, &copy_options)
```
Comment (jackcmay) [inline runtime/src/serde_snapshot.rs:317]: Do this in a different PR

File: runtime/src/serde_snapshot/tests.rs [hunk 14]
```diff
@@ -262,7 +262,7 @@ mod test_bank_serialize {
 
     // These some what long test harness is required to freeze the ABI of
     // Bank's serialization due to versioned nature
-    #[frozen_abi(digest = "ULV2jDndxR3JB677ayyjaamtAcZ24q75tCkHS2bKVoy")]
```
Comment (jackcmay) [inline runtime/src/serde_snapshot/tests.rs:265]: Same as earlier, I doubt any of these ABIs changed, check what Trent said, or if that's not what led to these we should identify what did

File: runtime/src/serde_snapshot.rs (hunk 13 above)
Comment (CriesofCarrots) [reply runtime/src/serde_snapshot.rs:317]: Ah, this looks familiar. Clippy was also suggesting this change on @t-nelson 's PR with the Cargo.lock craziness. I think it you follow his advice to reset that file, you'll be able to remove this change too.

File: programs/bpf_test_utils/Cargo.toml (hunk 10 above)
Comment (Lichtso) [reply programs/bpf_test_utils/Cargo.toml:2]: Tried it, but it results in a dependency cycle between solana-runtime and solana-sdk

File: programs/bpf_loader/src/lib.rs (hunk 7 above)
Comment (Lichtso) [reply programs/bpf_loader/src/lib.rs]: Do you mean the `.jit_compile()` or the `.unwrap()` ?

File: programs/bpf_loader/src/lib.rs (hunk 6 above)
Comment (Lichtso) [reply programs/bpf_loader/src/lib.rs:279]: Will do so in the JIT cache refactoring.

File: programs/bpf_test_utils/Cargo.toml (hunk 10 above)
Comment (mvines) [reply programs/bpf_test_utils/Cargo.toml:2]: I think this can just go directly into runtime.  We'll need these mocks from program-based tests as well.   It can't go into sdk because InvokeContext/etc are runtime-only things (which I think is a good spot for them).

I'd even like to see `bpf_test_utils` landed sooner, in a separate PR.  We'll want the mocks in 1.4 as well, so a separate PR will be easier to backporting

File: core/src/cluster_info.rs (hunk 3 above)
Comment (Lichtso) [reply core/src/cluster_info.rs]: Thx, solved all cases of `#[frozen_abi(digest = `

File: programs/bpf_loader/src/syscalls.rs (hunk 9 above)
Comment (Lichtso) [reply programs/bpf_loader/src/syscalls.rs:102]: That has to do with the refactoring of syscall functions and objects, how they are registered and bound. It was easier this way than to have a separate method for all combinations of these. But we can try to turn it back, once the JIT caching refactoring is done.

File: runtime/src/serde_snapshot.rs (hunk 13 above)
Comment (Lichtso) [reply runtime/src/serde_snapshot.rs]: Could be solved by reverting the dependencies, as you said.

File: programs/bpf_loader/src/lib.rs (hunk 7 above)
Comment (jackcmay) [reply programs/bpf_loader/src/lib.rs]: Both, but good point, replace this `unwrap` with a friendly error

File: programs/bpf_loader/src/syscalls.rs (hunk 9 above)
Comment (jackcmay) [reply programs/bpf_loader/src/syscalls.rs:102]: Looked like all the calls required a call to `hash_symbol_name`, that could just be done internally and pass a `&str` instead?

File: programs/bpf_loader/src/lib.rs (hunk 8 above)
Comment (Lichtso) [reply programs/bpf_loader/src/lib.rs:258]: There is still some shared code in the VM to setup the `memory_mapping` for the interpreter and the JIT. That is a reason to keep the VM constructor.

File: programs/bpf_test_utils/Cargo.toml (hunk 10 above)
Comment (Lichtso) [reply programs/bpf_test_utils/Cargo.toml]: `bpf_test_utils` is now a module in `solana_runtime`, should I copy that to an extra pull request and revert the API changes there?

Comment (jackcmay) [reply programs/bpf_test_utils/Cargo.toml]: Be cleaner if you broke it out

Comment (Lichtso) [reply programs/bpf_test_utils/Cargo.toml]: Done: https://github.com/solana-labs/solana/pull/13203

File: programs/bpf_loader/src/syscalls.rs (hunk 9 above)
Comment (Lichtso) [reply programs/bpf_loader/src/syscalls.rs:102]: Refactoring is done, and register by name is back: https://github.com/solana-labs/rbpf/pull/107

**Comment by mvines** [comment]
Backing this out in https://github.com/solana-labs/solana/pull/13286 due to CPI crashes I'm seeing in SPL land.  😞

**Comment by mvines** [comment]
STR:
1. Sync to the tip of master, and apply this PR again.  Sorry for the slight merge conflicts 🙇 
2. Clone https://github.com/mvines/solana-program-library/tree/dta
3. From your clone of (2), run:
```
$ ./patch.crates-io.sh <local path to monorepo from (1)>
$ cd default-token-account/program/
$ <local path to monorepo from (1)>/cargo-build-bpf   
$ cargo test sanity
```

You should get a panic like this:
```
thread 'solana-bank-forks-client' panicked at 'assertion failed: jit.offset + size <= jit.contents.len()', /rustc/04488afe34512aa4c33566eb16d8c912a3ae04f9/src/libstd/macros.rs:13:23
```

**Comment by Lichtso** [comment]
That means that the JIT didn't allocate enough memory for the compiled code. I will look into which program it is and why it produces so much code.

**Comment by mvines** [comment]
Thanks!  The program is the standard spl-token, https://explorer.solana.com/address/TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA

The shared object is also here: https://github.com/mvines/solana-program-library/blob/dta/default-token-account/program/tests/fixtures/spl_token.so
